src/librustc/infer/error_reporting/nice_region_error/trait_impl_difference.rs

```diff
@@ -46,14 +51,87 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
         None
     }
 
-    fn emit_err(&self, sp: Span, expected: Ty<'tcx>, found: Ty<'tcx>, impl_sp: Span) {
+    fn emit_err(&self, sp: Span, expected: Ty<'tcx>, found: Ty<'tcx>, trait_def_id: DefId) {
+        let tcx = self.tcx();
+        let trait_sp = self.tcx().def_span(trait_def_id);
         let mut err = self.tcx().sess.struct_span_err(
             sp,
             "`impl` item signature doesn't match `trait` item signature",
         );
-        err.note(&format!("expected `{:?}`\n   found `{:?}`", expected, found));
         err.span_label(sp, &format!("found {:?}", found));
-        err.span_label(impl_sp, &format!("expected {:?}", expected));
+        err.span_label(trait_sp, &format!("expected {:?}", expected));
+        let trait_fn_sig = tcx.fn_sig(trait_def_id);
+
+        struct AssocTypeFinder(FxHashSet<ty::ParamTy>);
+        impl<'tcx> ty::fold::TypeVisitor<'tcx> for AssocTypeFinder {
+            fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
+                debug!("assoc type finder ty {:?} {:?}", ty, ty.kind);
+                match ty.kind {
+                    ty::Param(param) => {
+                        self.0.insert(param);
+                    }
+                    _ => {}
+                }
+                ty.super_visit_with(self)
+            }
+        }
+        let mut visitor = AssocTypeFinder(FxHashSet::default());
+        trait_fn_sig.output().visit_with(&mut visitor);
+
+        if let Some(id) = tcx.hir().as_local_hir_id(trait_def_id) {
+            let parent_id = tcx.hir().get_parent_item(id);
+            let trait_item = tcx.hir().expect_item(parent_id);
+            if let hir::ItemKind::Trait(_, _, generics, _, _) =  &trait_item.kind {
+                for param_ty in visitor.0 {
```

Comment:
Iterating a `FxHashMap` --> non-determinism in errors?

Reply:
It's an `FxHashSet`, but the point stands. I think this is unlikely to cause any real issues in practice because we are just pointing at the spans, stderr tests do not care what order the labels got populated in.

Reply:
It's unlikely yes, but `.stderr` tests not caring is something that might not hold forever, and this is an error path, so why not just use an `FxIndexSet` instead?

Reply:
done